File: Session16/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n=len(s)
        # dp=[[False]*n]*n Wrong way to make 2d array
        dp = [[0 for i in range(n)] for j in range(n)]

        maxlen=1
        maxlenstpt=0
        
        for i in range(0,n):
            dp[i][i]=True
            if i<n-1 and s[i]==s[i+1]:
                maxlen=2
                maxlenstpt=i
                dp[i][i+1]=True
                
                
        
        for l in range(3,n+1):
            for st in range(0,n-l+1):
                e=st+l-1
                if s[st]==s[e] and dp[st+1][e-1]==True:
                    dp[st][e]=True
                    if l>maxlen:
                        maxlen=l
                        maxlenstpt=st
                        
        return s[maxlenstpt:maxlenstpt+maxlen]
    

File: Session16/5. Longest Palindromic Substring.java
class Solution {
    public String longestPalindrome(String s) {
        int n=s.length();
        int dp[][]=new int[n][n];
        int maxlen=1;
        int startind=0;
        
        for(int i=0;i<n;i++)
        {
            dp[i][i]=1;
            if(i+1<n && s.charAt(i)==s.charAt(i+1))
            {
                dp[i][i+1]=1;
                startind=i;
                maxlen=2;
            }
        }
        
        
        for(int l=3;l<=n;l++)
        {
            for(int i=0;i<n-l+1;i++)
            {
                int j=i+l-1;
                
                    if(s.charAt(i)==s.charAt(j) && dp[i+1][j-1]==1)
                    {
                        dp[i][j]=1;
                        if(l>maxlen)
                        {
                            maxlen=l;
                            startind=i;
                        }
                    }
            }
        }
        
        return s.substring(startind,startind+maxlen);
    }
}